File: web/lib/scheduled-item.ts
/**
 * `ScheduledItem` — the flattened value type that views and the timeline builder
 * consume (port of Swift `ScheduledItem` / `TimelineBlock` / `DayTimeline`).
 *
 * A `ScheduledItem` never touches persistence directly; the store projects the
 * stored models into these.
 */

import type { PastelToken } from "./palette";
import { freeTimeLayout, standardDayBounds } from "./freetime";
import { type TimeSlot, clamped, duration as slotDuration } from "./timeslot";
import { startOfDay } from "./time";

export type UniversityEventKind =
  | "lecture"
  | "practice"
  | "exam"
  | "holiday"
  | "vacation";

export type ItemKind =
  | { type: "university"; kind: UniversityEventKind }
  | { type: "work" }
  | { type: "sport" }
  | { type: "study" }
  | { type: "nap" }
  | { type: "custom" }
  | { type: "externalCalendar" };

export function isPersonalKind(kind: ItemKind): boolean {
  switch (kind.type) {
    case "university":
    case "externalCalendar":
      return false;
    default:
      return true;
  }
}

/** Non-colour accessibility differentiator (§45): a short noun. */
export function accessibilityCategory(kind: ItemKind): string {
  if (kind.type === "university") {
    switch (kind.kind) {
      case "lecture":
        return "Clase";
      case "practice":
        return "Práctica";
      case "exam":
        return "Examen";
      case "holiday":
        return "Festivo";
      case "vacation":
        return "Vacaciones";
    }
  }
  switch (kind.type) {
    case "work":
      return "Trabajo";
    case "sport":
      return "Deporte";
    case "study":
      return "Estudio";
    case "nap":
      return "Siesta";
    case "custom":
      return "Actividad";
    case "externalCalendar":
      return "Calendario";
    default:
      return "Actividad";
  }
}

export interface ScheduledItem {
  id: string;
  /** Compact label shown on the card (e.g. the subject code `"BI"`). */
  title: string;
  /** Optional line shown above the title, e.g. `"EXAMEN · BI"` (§25). */
  badge?: string | null;
  /** Full name for screen readers / detail views when `title` is an abbreviation. */
  fullTitle?: string | null;
  /** epoch ms */
  start: number;
  /** epoch ms */
  end: number;
  kind: ItemKind;
  palette: PastelToken;
  /** Lucide icon name. */
  symbolName: string;
  /** Stable key for a university instance, to omit/restore one class (§14). */
  instanceKey?: string | null;
  /** University lectures: can be lifted visually but never re-timed (§14). */
  isImmovable?: boolean;
  /** Locked against accidental drags until explicitly unlocked (§17). */
  isPinned?: boolean;
  /** Optional "done" mark for personal activities (§20). */
  isCompleted?: boolean;
}

export function itemSlot(item: ScheduledItem): TimeSlot {
  return { start: item.start, end: item.end };
}

export function itemDuration(item: ScheduledItem): number {
  return slotDuration(itemSlot(item));
}

export function isPersonal(item: ScheduledItem): boolean {
  return isPersonalKind(item.kind);
}

// MARK: - Timeline

export type TimelineBlock =
  | { kind: "item"; id: string; item: ScheduledItem }
  | { kind: "free"; id: string; slot: TimeSlot };

export function blockStart(b: TimelineBlock): number {
  return b.kind === "item" ? b.item.start : b.slot.start;
}
export function blockEnd(b: TimelineBlock): number {
  return b.kind === "item" ? b.item.end : b.slot.end;
}

export interface DayTimeline {
  /** epoch ms, start of day */
  date: number;
  /** Effective top of the timeline (09:00, or earlier if an activity beats it). */
  start: number;
  /** Bottom of the timeline (00:00 next day). */
  end: number;
  blocks: TimelineBlock[];
  /** A discreet note shown at the top on holidays / vacation days (§26). */
  dayNote?: string | null;
}

export function timelineItems(t: DayTimeline): ScheduledItem[] {
  return t.blocks.flatMap((b) => (b.kind === "item" ? [b.item] : []));
}

/**
 * Builds a laid-out day from an ordered list of items (already filtered for
 * omitted classes / holidays by the assembler).
 */
export function buildDayTimeline(
  date: Date,
  activities: ScheduledItem[],
  dayNote?: string | null,
): DayTimeline {
  const bounds = standardDayBounds(date);
  const layout = freeTimeLayout(
    activities.map(itemSlot),
    bounds,
  );

  const window: TimeSlot = { start: layout.effectiveStart, end: layout.end };
  const itemBlocks: TimelineBlock[] = activities.flatMap((item) => {
    const visible = clamped(itemSlot(item), window);
    if (!visible) return [];
    return [
      {
        kind: "item" as const,
        id: `item-${item.id}`,
        item: { ...item, start: visible.start, end: visible.end },
      },
    ];
  });
  const freeBlocks: TimelineBlock[] = layout.freeSlots.map((s) => ({
    kind: "free" as const,
    id: `free-${s.start}`,
    slot: s,
  }));

  const blocks = [...itemBlocks, ...freeBlocks].sort(
    (a, b) => blockStart(a) - blockStart(b),
  );

  return {
    date: startOfDay(date).getTime(),
    start: layout.effectiveStart,
    end: layout.end,
    blocks,
    dayNote: dayNote ?? null,
  };
}

File: apple/Horadia/Scheduling/ScheduledItem.swift
import Foundation

/// A concrete, placed block on a day's timeline — the value type that views and
/// the timeline builder consume. It is a flattened projection of the persistence
/// models (`ScheduledActivity`, `UniversityEvent`, routines, external calendar
/// events) so the UI never touches SwiftData directly (section 48).
struct ScheduledItem: Identifiable, Equatable {
    let id: UUID
    /// Compact label shown on the card (e.g. the subject code `"BI"`).
    var title: String
    /// Optional line shown above the title, e.g. `"EXAMEN · BI"` (section 25).
    var badge: String?
    /// Full name for VoiceOver / detail views, when `title` is an abbreviation.
    var fullTitle: String? = nil
    var start: Date
    var end: Date
    var kind: Kind
    var palette: PastelColor
    var symbolName: String

    /// Stable key for a university instance (`UniversityEventOverride.key`), used
    /// to omit/restore a single class on a single day (section 14). `nil` for
    /// personal activities.
    var instanceKey: String? = nil

    // Flags (sections 14, 17, 20)
    /// University lectures: can be lifted visually but never re-timed (section 14).
    var isImmovable: Bool = false
    /// Locked against accidental drags until explicitly unlocked (section 17).
    var isPinned: Bool = false
    /// Optional "done" mark for personal activities (section 20).
    var isCompleted: Bool = false

    var slot: TimeSlot { TimeSlot(start: start, end: end) }
    var duration: TimeInterval { slot.duration }

    /// `true` for anything the user may freely move / resize / delete.
    var isPersonal: Bool { kind.isPersonal }

    enum Kind: Equatable, Hashable {
        case university(UniversityEventKind)
        case work
        case sport
        case study
        case nap
        case custom
        case externalCalendar

        var isPersonal: Bool {
            switch self {
            case .university, .externalCalendar: return false
            case .work, .sport, .study, .nap, .custom: return true
            }
        }

        /// Non-colour accessibility differentiator (section 45): a short noun.
        var accessibilityCategory: String {
            switch self {
            case .university(.lecture): return "Clase"
            case .university(.practice): return "Práctica"
            case .university(.exam): return "Examen"
            case .university(.holiday): return "Festivo"
            case .university(.vacation): return "Vacaciones"
            case .work: return "Trabajo"
            case .sport: return "Deporte"
            case .study: return "Estudio"
            case .nap: return "Siesta"
            case .custom: return "Actividad"
            case .externalCalendar: return "Calendario"
            }
        }
    }
}

/// One row of a laid-out day: either a real item or a computed "Libre" gap.
enum TimelineBlock: Identifiable, Equatable {
    case item(ScheduledItem)
    case free(TimeSlot)

    var id: String {
        switch self {
        case .item(let item): return "item-\(item.id.uuidString)"
        case .free(let slot): return "free-\(slot.start.timeIntervalSinceReferenceDate)"
        }
    }

    var slot: TimeSlot {
        switch self {
        case .item(let item): return item.slot
        case .free(let slot): return slot
        }
    }

    var start: Date { slot.start }
    var end: Date { slot.end }
}

/// A fully laid-out single day, ready to render (section 5, 33).
struct DayTimeline: Equatable {
    var date: Date
    /// Effective top of the timeline (09:00, or earlier if an activity beats it).
    var start: Date
    /// Bottom of the timeline (00:00 next day).
    var end: Date
    var blocks: [TimelineBlock]
    /// A discreet note shown at the top on holidays / vacation days (section 26).
    var dayNote: String?

    var items: [ScheduledItem] {
        blocks.compactMap { if case .item(let i) = $0 { return i } else { return nil } }
    }

    static func build(
        date: Date,
        activities: [ScheduledItem],
        dayNote: String? = nil,
        calendar: Calendar = HoradiaCalendar.current
    ) -> DayTimeline {
        let bounds = FreeTimeEngine.DayBounds.standard(for: date, calendar: calendar)
        let layout = FreeTimeEngine.layout(
            activities: activities.map(\.slot),
            bounds: bounds
        )

        // Clamp item display to the day window so a midnight-crosser paints only
        // its portion of *this* day.
        let window = TimeSlot(start: layout.effectiveStart, end: layout.end)
        let itemBlocks: [TimelineBlock] = activities.compactMap { item in
            guard let visible = item.slot.clamped(to: window) else { return nil }
            var projected = item
            projected.start = visible.start
            projected.end = visible.end
            return .item(projected)
        }
        let freeBlocks = layout.freeSlots.map { TimelineBlock.free($0) }

        let blocks = (itemBlocks + freeBlocks).sorted { $0.start < $1.start }
        return DayTimeline(
            date: date,
            start: layout.effectiveStart,
            end: layout.end,
            blocks: blocks,
            dayNote: dayNote
        )
    }
}
